src/Npgsql/Internal/NpgsqlConnector.cs

```diff
@@ -782,13 +782,25 @@ async Task RawOpen(SslMode sslMode, NpgsqlTimeout timeout, bool async, Cancellat
 
                         ProvideClientCertificatesCallback?.Invoke(clientCertificates);
 
-                        var certificateValidationCallback = Settings.TrustServerCertificate
-                            ? SslTrustServerValidation
-                            : (Settings.RootCertificate ?? PostgresEnvironment.SslCertRoot ?? PostgresEnvironment.SslCertRootDefault) is { } certRootPath
-                                ? SslRootValidation(certRootPath)
-                                : UserCertificateValidationCallback is { } userValidation
-                                    ? userValidation
-                                    : SslDefaultValidation;
+                        RemoteCertificateValidationCallback? certificateValidationCallback;
+                        if (sslMode == SslMode.Prefer || sslMode == SslMode.Require && (Settings.TrustServerCertificate ?? true))
+                        {
+                            certificateValidationCallback = SslTrustServerValidation;
+                        }
+                        else if ((Settings.RootCertificate ?? PostgresEnvironment.SslCertRoot ?? PostgresEnvironment.SslCertRootDefault) is { } certRootPath)
+                        {
+                            certificateValidationCallback = SslRootValidation(certRootPath);
+                        }
+                        else if (UserCertificateValidationCallback is { } userValidation)
+                        {
+                            certificateValidationCallback = userValidation;
+                        }
+                        else
+                        {
+                            certificateValidationCallback = sslMode == SslMode.VerifyCA
+                                ? SslVerifyCAValidation
+                                : SslVerifyFullValidation;
+                        }
```

Comment:
Just making things slightly explicit/easier to follow:

```suggestion
                        else if (sslMode == SslMode.VerifyCA)
                        {
                            certificateValidationCallback = SslVerifyCAValidation;
                        }
                        else
                        {
                            Debug.Assert(sslMode == SslMode.VerifyFull);
                            certificateValidationCallback = SslVerifyFullValidation;
                        }
```